File: gis/src/commonMain/kotlin/jetbrains/gis/geoprotocol/GeoResponseBuilder.java
package jetbrains.gis.geoprotocol;

import jetbrains.datalore.base.projectionGeometry.GeoRectangle;
import jetbrains.gis.protocol.GeoResponse.AmbiguousGeoResponse;
import jetbrains.gis.protocol.GeoResponse.AmbiguousGeoResponse.AmbiguousFeature;
import jetbrains.gis.protocol.GeoResponse.AmbiguousGeoResponse.Namesake;
import jetbrains.gis.protocol.GeoResponse.AmbiguousGeoResponse.NamesakeParent;
import jetbrains.gis.protocol.GeoResponse.SuccessGeoResponse;
import jetbrains.gis.protocol.GeoResponse.SuccessGeoResponse.GeocodedFeature;
import jetbrains.ocelot.base.geometry.DoubleVector;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

import static com.google.common.collect.Streams.zip;
import static java.util.stream.Collectors.toList;

public final class GeoResponseBuilder {
  private GeoResponseBuilder() {
  }

  public static class SuccessResponseBuilder {
    private List<GeocodedFeature> myGeocodedFeatures = new ArrayList<>();
    private Optional<FeatureLevel> myFeatureLevel = Optional.empty();

    public SuccessResponseBuilder addGeocodedFeature(GeocodedFeature feature) {
      myGeocodedFeatures.add(feature);
      return this;
    }

    public SuccessResponseBuilder addGeocodedFeatures(List<GeocodedFeature> features) {
      features.forEach(this::addGeocodedFeature);
      return this;
    }

    public SuccessResponseBuilder setLevel(Optional<FeatureLevel> level) {
      myFeatureLevel = level;
      return this;
    }

    public SuccessGeoResponse build() {
      return new SuccessGeoResponse(myGeocodedFeatures, myFeatureLevel);
    }
  }

  public static class AmbiguousResponseBuilder {
    private List<AmbiguousFeature> myAmbiguousFeatures = new ArrayList<>();
    private Optional<FeatureLevel> myFeatureLevel;

    public AmbiguousResponseBuilder addAmbiguousFeature(AmbiguousFeature feature) {
      myAmbiguousFeatures.add(feature);
      return this;
    }

    public AmbiguousResponseBuilder addAmbiguousFeatures(List<AmbiguousFeature> features) {
      features.forEach(this::addAmbiguousFeature);
      return this;
    }

    public AmbiguousResponseBuilder setLevel(Optional<FeatureLevel> v) {
      myFeatureLevel = v;
      return this;
    }

    public AmbiguousGeoResponse build() {
      return new AmbiguousGeoResponse(myAmbiguousFeatures, myFeatureLevel);
    }
  }

  public static class GeocodedFeatureBuilder {
    private String query;
    private String id;
    private String name;
    private DoubleVector centroid;
    private GeoRectangle limit;
    private GeoRectangle position;
    private List<String> highlights = new ArrayList<>();
    private Geometry boundary;
    private List<GeoTile> tileGeometries = new ArrayList<>();

    public GeocodedFeatureBuilder setQuery(String v) {
      this.query = v;
      return this;
    }

    public GeocodedFeatureBuilder setId(String id) {
      this.id = id;
      return this;
    }

    public GeocodedFeatureBuilder setName(String name) {
      this.name = name;
      return this;
    }

    public GeocodedFeatureBuilder setBoundary(Geometry boundary) {
      this.boundary = boundary;
      return this;
    }

    public GeocodedFeatureBuilder addTile(GeoTile tile) {
      this.tileGeometries.add(tile);
      return this;
    }

    public GeocodedFeatureBuilder addTiles(List<GeoTile> tiles) {
      tiles.forEach(this::addTile);
      return this;
    }

    public GeocodedFeatureBuilder setCentroid(DoubleVector centroid) {
      this.centroid = centroid;
      return this;
    }

    public GeocodedFeatureBuilder setLimit(GeoRectangle limit) {
      this.limit = limit;
      return this;
    }

    public GeocodedFeatureBuilder setPosition(GeoRectangle position) {
      this.position = position;
      return this;
    }

    public GeocodedFeatureBuilder addHighlight(String v) {
      this.highlights.add(v);
      return this;
    }

    public GeocodedFeatureBuilder addHighlights(List<String> highlights) {
      highlights.forEach(this::addHighlight);
      return this;
    }

    public GeocodedFeature build() {
      return new GeocodedFeature(
          query,
          id,
          name,
          highlights,
          centroid,
          position,
          limit,
          boundary,
          tileGeometries
      );
    }
  }

  public static class AmbiguousFeatureBuilder {
    private String query;
    private int totalNamesakeCount;
    private List<Namesake> namesakeExamples = new ArrayList<>();

    public AmbiguousFeature build() {
      return new AmbiguousFeature(query, totalNamesakeCount, namesakeExamples);
    }

    public AmbiguousFeatureBuilder setQuery(String query) {
      this.query = query;
      return this;
    }

    public AmbiguousFeatureBuilder setTotalNamesakeCount(int totalNamesakeCount) {
      this.totalNamesakeCount = totalNamesakeCount;
      return this;
    }

    public AmbiguousFeatureBuilder addNamesakeExample(Namesake v) {
      this.namesakeExamples.add(v);
      return this;
    }

    public AmbiguousFeatureBuilder addNamesakeExamples(List<Namesake> namesakes) {
      namesakes.forEach(this::addNamesakeExample);
      return this;
    }
  }

  public static class NamesakeBuilder {
    private String name;
    private List<String> parentNames = new ArrayList<>();
    private List<FeatureLevel> parentLevels = new ArrayList<>();

    public Namesake build() {
      if (parentNames.size() != parentLevels.size()) {
        throw new IllegalStateException();
      }

      List<NamesakeParent> namesakeParents = zip(
          parentNames.stream(),
          parentLevels.stream(),
          NamesakeParent::new
      ).collect(toList());

      return new Namesake(name, namesakeParents);
    }

    public NamesakeBuilder setName(String name) {
      this.name = name;
      return this;
    }

    public NamesakeBuilder addParentName(String v) {
      parentNames.add(v);
      return this;
    }

    public NamesakeBuilder addParentLevel(FeatureLevel v) {
      parentLevels.add(v);
      return this;
    }

    public NamesakeBuilder addParent(String name, FeatureLevel level) {
      addParentName(name);
      addParentLevel(level);
      return this;
    }

    public NamesakeBuilder addParents(List<NamesakeParent> parents) {
      parents.forEach(parent -> addParent(parent.getName(), parent.getLevel()));
      return this;
    }
  }
}

File: gis/src/commonMain/kotlin/jetbrains/gis/geoprotocol/GeoResponseBuilder.kt
/*
 * Copyright (c) 2019. JetBrains s.r.o.
 * Use of this source code is governed by the MIT license that can be found in the LICENSE file.
 */

package jetbrains.gis.geoprotocol

import jetbrains.datalore.base.projectionGeometry.Generic
import jetbrains.datalore.base.projectionGeometry.Vec
import jetbrains.datalore.base.spatial.GeoRectangle
import jetbrains.gis.geoprotocol.GeoResponse.AmbiguousGeoResponse
import jetbrains.gis.geoprotocol.GeoResponse.AmbiguousGeoResponse.*
import jetbrains.gis.geoprotocol.GeoResponse.SuccessGeoResponse
import jetbrains.gis.geoprotocol.GeoResponse.SuccessGeoResponse.GeocodedFeature


object GeoResponseBuilder {

    class SuccessResponseBuilder {
        private var geocodedFeatures: ArrayList<GeocodedFeature> = ArrayList()
        private var featureLevel: FeatureLevel? = null

        fun addGeocodedFeature(v: GeocodedFeature) = apply { geocodedFeatures.add(v) }
        fun setLevel(v: FeatureLevel?) = apply { featureLevel = v }
        fun build() = SuccessGeoResponse(geocodedFeatures, featureLevel)
    }

    class AmbiguousResponseBuilder {
        private var ambiguousFeatures: ArrayList<AmbiguousFeature> = ArrayList()
        private var featureLevel: FeatureLevel? = null

        fun addAmbiguousFeature(v: AmbiguousFeature) = apply { ambiguousFeatures.add(v) }
        fun setLevel(v: FeatureLevel?) = apply { featureLevel = v }
        fun build() = AmbiguousGeoResponse(ambiguousFeatures, featureLevel)
    }

    class GeocodedFeatureBuilder() {
        private lateinit var query: String
        private lateinit var id: String
        private lateinit var name: String
        private var centroid: Vec<Generic>? = null
        private var limit: GeoRectangle? = null
        private var position: GeoRectangle? = null
        private var boundary: Boundary<Generic>? = null
        private var highlights: MutableList<String> = ArrayList()
        private var tileGeometries: ArrayList<GeoTile> = ArrayList()

        fun setQuery(v: String) = apply { query = v }
        fun setId(v: String) = apply { id = v }
        fun setName(v: String) = apply { name = v }
        fun setBoundary(v: Boundary<Generic>) = apply { boundary = v }
        fun setCentroid(v: Vec<Generic>) = apply { centroid = v }
        fun setLimit(v: GeoRectangle) = apply { limit = v }
        fun setPosition(v: GeoRectangle) = apply { position = v }
        fun addHighlight(v: String) = apply { highlights.add(v) }
        fun addTile(v: GeoTile) = apply { tileGeometries.add(v) }

        fun build(): GeocodedFeature {
            return GeocodedFeature(
                query,
                id,
                name,
                centroid,
                position,
                limit,
                boundary,
                highlights.ifEmpty { null },
                tileGeometries.ifEmpty { null }
            )
        }
    }

    class AmbiguousFeatureBuilder {
        private lateinit var query: String
        private var totalNamesakeCount: Int = 0
        private var namesakeExamples: ArrayList<Namesake> = ArrayList()

        fun setQuery(v: String) = apply { query = v }
        fun addNamesakeExample(v: Namesake) = apply { namesakeExamples.add(v) }
        fun setTotalNamesakeCount(v: Int) = apply { totalNamesakeCount = v }
        fun build() = AmbiguousFeature(query, totalNamesakeCount, namesakeExamples)
    }

    class NamesakeBuilder {
        private lateinit var name: String
        private val parentNames = ArrayList<String>()
        private val parentLevels = ArrayList<FeatureLevel>()

        fun setName(v: String) = apply { name = v }
        fun addParentName(v: String) = apply { parentNames.add(v) }
        fun addParentLevel(v: FeatureLevel) = apply { parentLevels.add(v) }

        fun build(): Namesake {
            if (parentNames.size != parentLevels.size) {
                throw IllegalStateException()
            }

            return Namesake(
                name,
                parentNames.zip(parentLevels, ::NamesakeParent).toList()
            )
        }
    }
}
